src/schema/schema_create.c

```diff
@@ -322,6 +322,60 @@ __wt_schema_index_source(WT_SESSION_IMPL *session,
 }
 
 /*
+ * __fill_index --
+ *	Fill the index from the current contents of the table.
+ */
+static int
+__fill_index(WT_SESSION_IMPL *session, WT_TABLE *table, const char *name)
+{
+	WT_DECL_RET;
+	WT_CURSOR *tcur, *icur, *child;
+	WT_CURSOR_INDEX *cindex;
+	WT_CURSOR_TABLE *ctable;
+	WT_INDEX *idx;
+	WT_SESSION *sess;
+
+	sess = (WT_SESSION *)session;
+	tcur = NULL;
+	icur = NULL;
+	WT_RET(__wt_schema_open_colgroups(session, table));
+
+	/*
+	 * If the column groups have not been completely created,
+	 * there cannot be data inserted yet, and we're done.
+	 */
+	if (!table->cg_complete)
+		return (0);
+
+	WT_ERR(sess->open_cursor(sess, table->name, NULL, "readonly", &tcur));
+	WT_ERR(sess->open_cursor(sess, name, NULL, NULL, &icur));
+
+	ctable = (WT_CURSOR_TABLE *)tcur;
+	cindex = (WT_CURSOR_INDEX *)icur;
+	child = cindex->child;
+	idx = cindex->index;
+
+	while ((ret = tcur->next(tcur)) == 0) {
+		/*
+		 * For LSM indices, we must drop locks so that the
+		 * LSM worker thread can grab the schema lock during
+		 * a switch.
+		 */
+		WT_WITHOUT_LOCKS(session,
+		    ret = __wt_apply_single_idx(session, idx,
+		    child, ctable, child->insert));
```

Comment:
@michaelcahill, I think you are right to question its safety.  Can a checkpoint see a partially-completed index?  Probably, and I guess I would need to ask you about recovery implications!  I think dropping the lock probably has various issues we don't want to explore... 

I took another approach in this pull request: #2127 .  Basically the same as this branch, except for this lock-dropping code.  Instead, we tell LSM that we're holding the schema lock on its behalf (so it doesn't try to take it).  I don't know enough about LSM to know whether this is safe: the worker thread (which is the one that needs the lock) is asynchronous to the application thread, but it seems like file splits (where the lock is needed) are coordinated to be synchronous.  Is this too onerous or fragile?

The only other alternative I can think of is telling LSM not to split while we're filling.  Since an index being filled can be arbitrary size, this seems like it could have pretty bad performance implications for big cases.  But maybe it's okay -- how often do you need LSM indices that are dynamically created?

I suppose we could also punt and disallow dynamic index creation for LSM indices, period.

@agorrod, if you're around, perhaps you can weigh in.